src/cert.rs

```diff
@@ -145,13 +135,19 @@ fn version3(input: &mut untrusted::Reader) -> Result<(), Error> {
     )
 }
 
-pub(crate) fn certificate_serial_number(input: &mut untrusted::Reader) -> Result<(), Error> {
+pub(crate) fn lenient_certificate_serial_number(input: &mut untrusted::Reader) -> Result<(), Error> {
     // https://tools.ietf.org/html/rfc5280#section-4.1.2.2:
     // * Conforming CAs MUST NOT use serialNumber values longer than 20 octets."
     // * "The serial number MUST be a positive integer [...]"
-
-    let value = der::positive_integer(input)?;
-    if value.big_endian_without_leading_zero().len() > 20 {
+    //
+    // However, we don't enforce these constraints, as there are widely-deployed trust anchors
+    // and many X.509 implementations in common use that violate these constraints. This is called
+    // out by the same section of RFC 5280 as cited above:
+    //   Note: Non-conforming CAs may issue certificates with serial numbers
+    //   that are negative or zero.  Certificate users SHOULD be prepared to
+    //   gracefully handle such certificates.
+    let value = der::expect_tag_and_get_value(input, Tag::Integer)?;
+    if value.len() > 20 {
```

Comment:
Makes sense to me.

Style suggestion (not necessarily idiomatic since not everyone likes using `match` for `bool`):

```rust
match value.len() <= 20 {
    true => Ok(()),
    false => Err(Error::BadDer),
}
```

(I like this because for single-line match arms, this is more compact than the `if`/`else` version, and it lays out the match arms right next to each other which I find makes it easier to follow.)

(Alternatively, you could use an `if value.len() > 20 { .. } else { .. }` as the last expression, avoiding the explicit `return` statement which would definitely be more idiomatic than what you've got now.)

Reply:
> Style suggestion (not necessarily idiomatic since not everyone likes using match for bool)

I like that suggestion :+1:  Applied in my latest force push (combined with a small `fmt` fix to address the `cargo fmt` check failure).